### PR TITLE
Add trace::set_trace_information method

### PR DESCRIPTION
**Pull request by pierricgimmig**
Some providers need to be configured through ETW's TraceSetInformation
function which we wrap inside trace::set_trace_information to maintain
encapsulation of the trace handle.

Tests: Setup sampling profiling. All tests pass.

See discussion in #171

### CODE REVIEW

File: krabs/krabs/etw.hpp [hunk 1]
```diff
@@ -77,6 +77,16 @@ namespace krabs { namespace details {
          */
         EVENT_TRACE_PROPERTIES query();
 
+        /**
+         * <summary>
+         * Configures the ETW trace session settings.
+         * </summary>
```
Comment (swannman) [inline krabs/krabs/etw.hpp:83]: let's update the comment with a link to https://docs.microsoft.com/en-us/windows/win32/api/evntrace/nf-evntrace-tracesetinformation so that users know how to configure these parameters

File: krabs/krabs/etw.hpp [hunk 2]
```diff
@@ -77,6 +77,16 @@ namespace krabs { namespace details {
...
+         */
+        void set_trace_information(
+            TRACE_INFO_CLASS informationClass,
```
Comment (swannman) [inline krabs/krabs/etw.hpp:86]: let's use snake_case here to match the style of the surrounding code: `information_class`, `trace_information`, `information_length`

File: krabs/krabs/etw.hpp [hunk 3]
```diff
@@ -168,7 +178,9 @@ namespace krabs { namespace details {
     template <typename T>
     EVENT_TRACE_LOGFILE trace_manager<T>::open()
     {
-        register_trace();
+        if (trace_.registrationHandle_ == INVALID_PROCESSTRACE_HANDLE) {
```
Comment (swannman) [inline krabs/krabs/etw.hpp:181]: is this change needed?

File: krabs/krabs/etw.hpp [hunk 4]
```diff
@@ -185,6 +197,25 @@ namespace krabs { namespace details {
         return query_trace();
     }
 
+    template <typename T>
+    void trace_manager<T>::set_trace_information(
+        TRACE_INFO_CLASS informationClass,
+        PVOID traceInformation,
+        ULONG informationLength)
+    {
+        if (trace_.registrationHandle_ == INVALID_PROCESSTRACE_HANDLE) {
```
Comment (swannman) [inline krabs/krabs/etw.hpp:206]: is this check and call needed? my preference would be to wrap TraceSetInformation as closely as possible without doing extra work on behalf of the caller, like calling `register_trace` if they haven't yet

File: krabs/krabs/etw.hpp [hunk 5]
```diff
@@ -185,6 +197,25 @@ namespace krabs { namespace details {
         return query_trace();
     }
 
+    template <typename T>
+    void trace_manager<T>::set_trace_information(
+        TRACE_INFO_CLASS informationClass,
```
Comment (swannman) [inline krabs/krabs/etw.hpp:202]: let's use snake_case for these three parameter names per the comment in etw.hpp

File: krabs/krabs/trace.hpp [hunk 6]
```diff
@@ -123,6 +123,30 @@ namespace krabs {
...
+         *    event_id.EventGuid = krabs::guids::perf_info;
+         *    event_id.Type = 46; // SampleProfile
+         *    trace_.set_trace_information(TraceStackTracingInfo, &event_id, sizeof(STACK_TRACING_EVENT_ID));
+         *    krabs::kernel_provider stack_walk_provider(EVENT_TRACE_FLAG_PROFILE, krabs::guids::stack_walk);
+         *    trace_.enable(stack_walk_provider);
+         *    trace.start();
+         * </example>
+         */
+        void set_trace_information(
+            TRACE_INFO_CLASS informationClass,
```
Comment (swannman) [inline krabs/krabs/trace.hpp:146]: let's use snake_case for these parameter names

File: krabs/krabs/trace.hpp [hunk 7]
```diff
@@ -319,6 +343,16 @@ namespace krabs {
         properties_.LogFileMode = properties->LogFileMode;
     }
 
+    template <typename T>
+    void trace<T>::set_trace_information(
+        TRACE_INFO_CLASS informationClass,
```
Comment (swannman) [inline krabs/krabs/trace.hpp:348]: let's use snake_case for these parameter names

**Review by swannman — CHANGES_REQUESTED** [review]
Looking good! Added a few minor observations.

File: krabs/krabs/etw.hpp (hunk 4 above)
Comment (pierricgimmig) [reply krabs/krabs/etw.hpp:207]: 'set_trace_information` needs the session handle, which is currently assigned by the blocking call to 'trace::start()'. This is impractical as we call 'trace::start()' in its own thread. We would then have to somehow synchronize with the assignment of the session handle before safely calling 'set_trace_information'. At the root, the fact 'trace::start' is responsible for both calls to 'StartTrace' and 'OpenTrace' is the issue here. The proposed change allows to split those calls if needed by calling `StartTrace`, which assigns the session handle, before the blocking call to 'trace::start()'. Am I missing something?

File: krabs/krabs/etw.hpp (hunk 5 above)
Comment (pierricgimmig) [reply krabs/krabs/etw.hpp]: Done.

File: krabs/krabs/trace.hpp (hunk 6 above)
Comment (pierricgimmig) [reply krabs/krabs/trace.hpp]: Done.

File: krabs/krabs/trace.hpp (hunk 7 above)
Comment (pierricgimmig) [reply krabs/krabs/trace.hpp]: Done.

File: krabs/krabs/etw.hpp (hunk 2 above)
Comment (pierricgimmig) [reply krabs/krabs/etw.hpp]: Done.

File: krabs/krabs/etw.hpp (hunk 1 above)
Comment (pierricgimmig) [reply krabs/krabs/etw.hpp:84]: Done.

File: krabs/krabs/etw.hpp (hunk 4 above)
Comment (swannman) [reply krabs/krabs/etw.hpp:207]: Got it! Rather than adding a call to `register_trace` here, would this sequence of calls give you the outcome you're looking for?

```
krabs::trace trace;
krabs::guid id(L"{A0C1853B-5C40-4B15-8766-3CF1C58F985A}");
provider<> powershell(id);
trace.enable(powershell);
trace.open(); // register_trace is called in here
trace.set_trace_information(...); // new
trace.process(); // this blocks
```

Comment (pierricgimmig) [reply krabs/krabs/etw.hpp]: Ah yes, I was indeed missing something :-). Thanks, yep this is perfect, I verified that it works. I removed the checks and modified the function comment for `set_trace_information`.

**Review by swannman — APPROVED** [review]
Looks great!

**Comment by swannman** [comment]
This change will be included in the 4.2.1 nuget version which I am publishing today.

**Comment by pierricgimmig** [comment]
> This change will be included in the 4.2.1 nuget version which I am publishing today.

Awesome, thanks a lot for your help, @swannman !
